File: 11-Diameter-of-Binary-Tree/diameterOfBinaryTree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	leftDepth := maxDepth(root.Left)
	rightDepth := maxDepth(root.Right)
	return int(math.Max(float64(leftDepth), float64(rightDepth))) + 1
}

func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}
	middle := maxDepth(root.Left) + maxDepth(root.Right)
	left := maxDepth(root.Left)
	right := maxDepth(root.Right)
	return int(math.Max(float64(middle), float64(math.Max(float64(left), float64(right)))))
}

func main() {

}

File: 11-Diameter-of-Binary-Tree/diameterOfBinaryTree.py
class TreeNode:
    def __init__(self) -> None:
        self.values = 0
        self.left = None
        self.right = None


def maxDepth(root: TreeNode):
    if root == None:
        return 0
    leftDepth = maxDepth(root.left)
    rightDepth = maxDepth(root.right)
    return max(leftDepth, rightDepth) + 1


def diameterOfBinaryTree(root: TreeNode):
    if root == None:
        return 0
    middle = maxDepth(root.left) + maxDepth(root.right)
    leftDiameter = diameterOfBinaryTree(root.left)
    rightDiameter = diameterOfBinaryTree(root.right)
    return max(middle, leftDiameter, rightDiameter)
